monai/data/utils.py

```diff
@@ -315,63 +315,38 @@ def decollate_batch(data: Union[dict, list, torch.Tensor], batch_size: Optional[
 
-        batch_data = torch.rand((2,1,10,10)
+        batch_data = torch.rand((2,1,10,10))
         out = decollate_batch(batch_data)
         print(out[0])
         >>> tensor([[[4.3549e-01...43e-01]]])
 
     Args:
-        data: data to be de-collated.
-        batch_size: number of batches in data. If `None` is passed, try to figure out batch size.
-    """
-
-    def torch_to_single(d: torch.Tensor):
-        """If input is a torch.Tensor with only 1 element, return just the element."""
-        return d if d.numel() > 1 else d.item()
-
-    def decollate(data: Any, idx: int):
-        """Recursively de-collate."""
-        if isinstance(data, dict):
-            return {k: decollate(v, idx) for k, v in data.items()}
-        if isinstance(data, torch.Tensor):
-            out = data[idx]
-            return torch_to_single(out)
-        if isinstance(data, list):
-            if len(data) == 0:
-                return data
-            if isinstance(data[0], torch.Tensor):
-                return [torch_to_single(d[idx]) for d in data]
-            if issequenceiterable(data[0]):
-                return [decollate(d, idx) for d in data]
-            return data[idx]
-        raise TypeError(f"Not sure how to de-collate type: {type(data)}")
-
-    def _detect_batch_size(batch_data):
-        for v in batch_data:
-            if isinstance(v, torch.Tensor):
-                return v.shape[0]
-        warnings.warn("batch_data is not a sequence of tensors in decollate, use `len(batch_data[0])` directly.")
-        return len(batch_data[0])
-
-    result: List[Any]
-    if isinstance(data, dict):
-        batch_size = _detect_batch_size(batch_data=data.values()) if batch_size is None else batch_size
-        result = [{key: decollate(data[key], idx) for key in data.keys()} for idx in range(batch_size)]
-    elif isinstance(data, list):
-        batch_size = _detect_batch_size(batch_data=data) if batch_size is None else batch_size
-        result = [[decollate(d, idx) for d in data] for idx in range(batch_size)]
-    elif isinstance(data, torch.Tensor):
-        batch_size = data.shape[0]
-        result = [data[idx] for idx in range(batch_size)]
-    else:
-        raise NotImplementedError("Only currently implemented for dictionary, list or Tensor data.")
-
-    return result
+        batch: data to be de-collated.
+        detach: whether to detach the tensors. Scalars tensors will be detached into number types
+            instead of torch tensors.
+    """
+    if isinstance(batch, torch.Tensor):
+        if detach:
+            batch = batch.detach()
+        if batch.ndim == 0 and detach:
+            return batch.item()
+        out_list = torch.unbind(batch, dim=0)
+        if out_list[0].ndim == 0 and detach:
+            return [t.item() for t in out_list]
+        return list(out_list)
+    if isinstance(batch, Mapping):
+        _dict_list = {key: decollate_batch(batch[key]) for key in batch}
```

Comment:
Should we add `detach=detach` in the recursive call?

Reply:
that's a good one, I'll update soon